src/asciidocParser.ts

```diff
@@ -172,6 +173,14 @@ export class AsciidocParser {
     }
 
     try {
+      const workspaceFolder = vscode.workspace.getWorkspaceFolder(doc.uri)
+      if (workspaceFolder !== undefined) {
+        const asciiDoctorConfig = vscode.Uri.joinPath(workspaceFolder.uri, '.asciidoctorconfig')
+        if (fs.existsSync(asciiDoctorConfig.fsPath)) {
+          const asciiDoctorConfigContent = await vscode.workspace.fs.readFile(asciiDoctorConfig)
+          text = asciiDoctorConfigContent + '\n' + text
```

Comment:
@ahus1 could you please confirm that this how the `.asciidoctorconfig` content should be used? (i.e., the content is prepended in the AsciiDoc document)

Reply:
Hi @Mogztter - I summarized the handling of the `.asciidoctorconfig` file on this page: https://intellij-asciidoc-plugin.ahus1.de/docs/users-guide/features/advanced/asciidoctorconfig-file.html

What I see in this PR (if I read it correctly), is that this picks a single file in the workspace folder, and prepends it. 

This is a start, still it lacks some of the features used in the IntelliJ plugin. Maybe the additional functionality can be split into futher PRs.

* The file can also be named `.asciidoctorconfig.adoc`, as this enables preprocessor macros if the file would be used as a standard AsciiDoctor include.
* The file can be in any folder, and then it affects all files in its subfolders. So you want to search in all folder leading to the current folder, and they are cascading.
* I found that it would be better to add two newlines here, to avoid problems with stick-to-together-content (won't happen with just attributes, could happen with other content)
* At the beginning of each of the files, I append `:asciidoctorconfigdir: <absolute path filename>`, this comes in handy to setup for example an `imagesdir`. 
* When just prepending the content, I found that the line numbers rendered in warnings will be off by the number of lines prepended. To work around that, I added a preprocessor that calls `reader.push_include`, that won't affect the line numbers.

Code references:

* https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/64707c7632e1c5e6efc34100dc3cb2dd0bab4ea6/src/main/java/org/asciidoc/intellij/AsciiDocWrapper.java#L710-L745
* https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/main/src/main/java/org/asciidoc/intellij/asciidoc/PrependConfig.java

Reply:
> The file can also be named .asciidoctorconfig.adoc, as this enables preprocessor macros if the file would be used as a standard AsciiDoctor include.

I don't understand why `.asciidoctorconfig` or `.asciidoctorconfig.adoc` would be treated differently since we prepend the content of the file. In other words, Asciidoctor is not aware of the file name, no?

> The file can be in any folder, and then it affects all files in its subfolders. So you want to search in all folder leading to the current folder, and they are cascading.


```
app/
├─ .asciidoctorconfig
├─ public/
│  ├─ .asciidoctorconfig
│  ├─ foo.adoc
├─ src/
│  ├─ .asciidoctorconfig
│  ├─ bar.adoc
├─ baz.adoc
```

IIUC, `foo.adoc` will include the content of `app/.asciidoctorconfig` and the content of `public/.asciidoctorconfig` (in that order). Since the content of `public/.asciidoctorconfig` if added after the content of `app/.asciidoctorconfig` it will have a higher precedence?

Reply:
Thanks for all these details!

> This is a start, still it lacks some of the features used in the IntelliJ plugin. Maybe the additional functionality can be split into futher PRs.

This is my plan, especially the time that i have to work on the topic. i would like to provide a almost minimal but functional scope first. Objective: provide tested and documented feature for '.asciidoctorconfig' at the root of the workspace without major drawback.

> I found that it would be better to add two newlines here, to avoid problems with stick-to-together-content (won't happen with just attributes, could happen with other content)

Added

> At the beginning of each of the files, I append :asciidoctorconfigdir: <absolute path filename>, this comes in handy to setup for example an imagesdir.

Added

> When just prepending the content, I found that the line numbers rendered in warnings will be off by the number of lines prepended. To work around that, I added a preprocessor that calls reader.push_include, that won't affect the line numbers.

I found that it is added here https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/95bb8f7cb1e1d0f935f71d4a7e97ba06abe87400/src/main/java/org/asciidoc/intellij/asciidoc/PrependConfig.java#L23 but i'm a bit lost how to register a preprocessor in this code fo rnow. Will continue investigation as it seems required even for a first iteration

> The file can also be named .asciidoctorconfig.adoc, as this enables preprocessor macros if the file would be used as a standard AsciiDoctor include.
> The file can be in any folder, and then it affects all files in its subfolders. So you want to search in all folder leading to the current folder, and they are cascading.

i think these two can be done in separated PRs.

Reply:
+1 for keeping this PR small an concise!

> When just prepending the content, I found that the line numbers rendered in warnings will be off by the number of lines prepended.

I'm not familiar with the VS Code plugin if it (and where) it shows any warnings and errors generated during processing of the preview. If it doesn't this is not needed for the PR. Even if it does, it might still be OK to have this in a future PR IMHO. So I suggest to not see this as a blocker.

Reply:
> I don't understand why .asciidoctorconfig or .asciidoctorconfig.adoc would be treated differently since we prepend the content of the file. In other words, Asciidoctor is not aware of the file name, no?

This makes no different to the IDE as you described. It only makes a different if someone wants to use the file as a standard include in Asciidoctor: Asciidoctor will only process preprocessor macros on includes which have a well-known AsciiDoc extension like ".adoc". 

> IIUC, foo.adoc will include the content of app/.asciidoctorconfig and the content of public/.asciidoctorconfig (in that order). Since the content of public/.asciidoctorconfig if added after the content of app/.asciidoctorconfig it will have a higher precedence?

Sorry for not responding earlier to this: The file higher up in the hierarchy has the highest precedence, wich the file in the root of the project having the highest one. Looking at the final document with all the prepended information. it starts with the file with the highest precedence / the one from the root, and then continues with the files in the folders up to the file wich is pre-rendered.
This is also the way `.editorconfig` works.

Reply:
> This makes no difference to the IDE as you described. It only makes a different if someone wants to use the file as a standard include in Asciidoctor: Asciidoctor will only process preprocessor macros on includes which have a well-known AsciiDoc extension like ".adoc".

👍🏻 

> Sorry for not responding earlier to this: The file higher up in the hierarchy has the highest precedence, wich the file in the root of the project having the highest one.
> Looking at the final document with all the prepended information. it starts with the file with the highest precedence / the one from the root, and then continues with the files in the folders up to the file wich is pre-rendered.
This is also the way .editorconfig works.

I think we don't have the same definition of higher precedence 😄
But as far as I understand, the `.asciidoctorconfig` file closer to the file takes precedence (i.e., for instance, it will override an attribute defined at the root).

> EditorConfig files are read top to bottom and the most recent rules found take precedence. Properties from matching EditorConfig sections are applied in the order they were read, so properties in closer files take precedence.

In our case, we include the most distant `.asciidoctorconfig` first so that closer `.asciidoctorconfig` fille take precedence.

Reply:
> I think we don't have the same definition of higher precedence

Thanks @Mogztter - you're right and I'm wrong. I stand corrected.